app/code/Magento/User/Model/ResourceModel/User/Collection.php

```diff
@@ -22,4 +22,18 @@ protected function _construct()
     {
         $this->_init(\Magento\User\Model\User::class, \Magento\User\Model\ResourceModel\User::class);
     }
+
+    public function _initSelect()
+    {
+        parent::_initSelect();
+        $this->getSelect()->join(
+            ['user_role' => $this->getTable('authorization_role')],
+            'main_table.user_id = user_role.user_id',
+            []
+        )->join(
+            ['detail_role' => $this->getTable('authorization_role')],
+            'user_role.parent_id = detail_role.role_id',
+            ['role_name']
```

Comment:
This will affect every user collection instantiated, isn't it?

Something like https://magento.stackexchange.com/a/124204, a separate collection just for grid, looks like a better option. Not sure if there is to way do it in more declarative way.

Just my 2 cents, before proceeding please wait for somebody to actually confirm such feature fits well into core.

Reply:
Yes the JOINs will happen on every collection load. This was based on the sales rule grid

https://github.com/magento/magento2/blob/75155ae0d42ce33bc91d7f42c19ee35f1f083d58/app/code/Magento/SalesRule/view/adminhtml/layout/sales_rule_promo_quote_index.xml#L15

https://github.com/magento/magento2/blob/develop/app/code/Magento/SalesRule/Model/ResourceModel/Rule/Quote/Collection.php#L10

https://github.com/magento/magento2/blob/develop/app/code/Magento/SalesRule/Model/ResourceModel/Rule/Collection.php#L279-L288

Not sure what the concern would be. The performance cost should be negligible.

Reply:
I searched for the usages of this Collection, it is not used in any performance-critical flows. Join should be fine here.

Reply:
@vrann glad to know, thanks! Hope you checked EE/B2B codebases as well.

Is it correct that such approach with `_initSelect` is up-to-date? Like, there is no declarative way to declare join for the grid etc.

Reply:
@vrann @orlangur also note that there are indexes on both the columns used in the JOIN so I really wouldn't be worried about perf.